docs/basic-tutorial.md

```diff
@@ -0,0 +1,685 @@
+    --grpc-swift_out=Sources/Examples/RouteGuide/Model
+```
+
+We invoke the protocol buffer compiler `protoc` with the path to our service
+definition `route_guide.proto` as well as specifying the path to search for
+imports. We then specify the path to the [Swift Protobuf][swift-protobuf] plugin
+and any options. In our case the generated code is in a separate module to the
+client and server, so the generated code must have `Public` visibility. We then
+specify the directory into which the generated messages should be written. The
+remainder of the arguments are very similar but pertain to the generation of the
+service code and use the `protoc-gen-grpc-swift` plugin.
+
+### Creating the server
+
+First let's look at how we create a `RouteGuide` server. If you're only
+interested in creating gRPC clients, you can skip this section and go straight
+to [Creating the client](#client) (though you might find it interesting
+anyway!).
+
+There are two parts to making our `RouteGuide` service do its job:
+
+- Implementing the service protocol generated from our service definition: doing
+  the actual "work" of our service.
+- Running a gRPC server to listen for requests from clients and return the
+  service responses.
+
+You can find our example `RouteGuide` provider in
+[grpc-swift/Sources/Examples/RouteGuide/Server/RouteGuideProvider.swift][routeguide-provider].
+Let's take a closer look at how it works.
+
+#### Implementing RouteGuide
+
+As you can see, our server has a `RouteGuideProvider` class that extends the
+generated `Routeguide_RouteGuideProvider` protocol:
+
+```swift
+class RouteGuideProvider: Routeguide_RouteGuideProvider {
+...
+}
+```
+
+#### Simple RPC
+
+`RouteGuideProvider` implements all our service methods. Let's
+look at the simplest type first, `GetFeature`, which just gets a `Point` from
+the client and returns the corresponding feature information from its database
+in a `Feature`.
+
+```swift
+/// A simple RPC.
+///
+/// Obtains the feature at a given position.
+///
+/// A feature with an empty name is returned if there's no feature at the given position.
+func getFeature(
+  request point: Routeguide_Point,
+  context: StatusOnlyCallContext
+) -> EventLoopFuture<Routeguide_Feature> {
+  return context.eventLoop.makeSucceededFuture(self.checkFeature(at: point))
+}
+
+...
+
+/// Returns a feature at the given location or an unnamed feature if none exist at that location.
+private func checkFeature(
+  at location: Routeguide_Point
+) -> Routeguide_Feature {
+  return self.features.first(where: {
+    return $0.location.latitude == location.latitude
+      && $0.location.longitude == location.longitude
+  }) ?? Routeguide_Feature.with {  // No feature was found: return an unnamed feature.
+    $0.name = ""
+    $0.location = location
+  }
+}
+```
+
+`getFeature()` takes two parameters:
+
+- `Routeguide_Point`: the request
+- `StatusOnlyCallContext`: a context which exposes status and trailing metadata
+  objects.
```

Comment:
s/objects/promises that you can fulfill/?

Reply:
They're not `EventLoopPromise`s: "objects that you can update"?

Reply:
Sorry, my mistake. How about "a context which exposes status and trailing metadata
  fields that you can change if needed."?